miio/gateway.py

```diff
@@ -536,38 +513,216 @@ class SubDevice:
+            return self.gw.send(command, arguments, extra_parameters={"sid": self.sid})
+        except Exception as ex:
+            _LOGGER.error(
+                "Got an exception while sending command '%s' with arguments '%s': %s", command, str(arguments), ex, exc_info=True
+            )
+            return None
+
+    @command(click.argument("property"))
+    def get_subdevice_prop(self, property):
+        """Get the value of a property of the subdevice."""
+        try:
+            return self.gw.send("get_device_prop", [self.sid, property])
         except Exception as ex:
             _LOGGER.debug(
-                "Got an exception while fetching fw_ver: %s", ex, exc_info=True
+                "Got an exception while fetching property %s: %s", property, ex, exc_info=True
             )
             return None
 
-    def __repr__(self):
-        return "<Subdevice %s: %s fw: %s bat: %s>" % (
-            self.type,
-            self.sid,
-            self.get_firmware_version(),
-            self.battery(),
-        )
+    @command(click.argument("properties", nargs=-1))
+    def get_subdevice_prop_exp(self, properties):
+        """Get the value of a bunch of properties of the subdevice."""
+        try:
+            return self.gw.send("get_device_prop_exp", [[self.sid] + list(properties)]).pop()
+        except Exception as ex:
+            _LOGGER.debug(
+                "Got an exception while fetching properties %s: %s", properties, ex, exc_info=True
+            )
+            return None
+
+    @command(click.argument("property"), click.argument("value"))
+    def set_subdevice_prop(self, property, value):
+        """Set a device property of the subdevice."""
+        try:
+            return self.gw.send("set_device_prop", {"sid": self.sid, property: value})
+        except Exception as ex:
+            _LOGGER.debug(
+                "Got an exception while setting propertie %s to value %s: %s", property, str(value), ex, exc_info=True
+            )
+            return None
+
+    @command()
+    def unpair(self):
+        return self.subdevice_send("remove_device")
+
+    @command()
+    def get_battery(self):
+        self._battery = self.subdevice_send("get_battery").pop()
+        return self._battery
+
+    @command()
+    def get_firmware_version(self) -> Optional[int]:
+        """Returns firmware version"""
+        return self.get_subdevice_prop("fw_ver").pop()
+
+
+class AqaraHT(SubDevice):
+    _temperature = None
+    _humidity = None
+    _pressure = None
+
+    @property
+    def temperature(self):
+        """Return the temperature in degrees celsius"""
+        return self._temperature
+
+    @property
+    def humidity(self):
+        """Return the humidity in %"""
+        return self._humidity
+
+    @property
+    def pressure(self):
+        """Return the pressure in hPa"""
+        return self._pressure
+
+    @command()
+    def update(self):
```

Comment:
Wrt. `@command` inside subdevices, have you thought how are these supposed to be exposed to the cli?

Reply:
Honestly I don't know how the cli works...

In python it would work like this:

from miio import (gateway)
gateway = gateway.Gateway("192.168.1.IP", 'token')
devices = gateway.discover_devices()

Now select the sub device you want to control, for instance the first discovered device:

Aqara1 = devices[0]
Aqara1.update()
print(Aqara1.temperature)

Reply:
The cli collects all methods of `Device`-derived class marked with `@command`, and adds them to `click.group` for that sub command. 

Unfortunately I don't have a clean solution either for this at the moment, but the aim would be to gain access to the group object of the parent class, which can be extended by adding new sub-groups (`zigbee`, `light`, ..) with their own commands.

Reply:
To add to that, the sub devices should not derive from `Device`, but we need to find a nicer way to enable cli access for those. As you mentioned already at https://github.com/rytilahti/python-miio/pull/700#issuecomment-632600444, having these initialized separately as Devices will cause problems in the long run..

So until we find a solution, I think we will simply have to leave these unexposed (or add some sort of simple interface to the gateway class to print out the details of known sub-devices. Maybe it will make sense to add a property to get the available information as dict for downstream users? Something like:
```
@property
def properties():
    return {"temperature": self._temperature, "humidity": self._humidity, "pressure": self._pressure}
```

Reply:
I now have the Gateway, GatewayAlarm, GatewayRadio, GatewayZigbee and GatewayLight working in cli, although not in the most beautifull way, but for the python implementation nothing changed and still only one device is initialized.

For subdevices things get dificult very soon......
I would like to leave that for someone else better experianced with cli and click

Reply:
Okay, I can fetch my gateway and a test device and try at least to expose the property values for the cli (after we get this PR merged).

Reply:
That would be very nice

Reply:
will clean up those parms now